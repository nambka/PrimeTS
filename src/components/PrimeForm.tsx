import React, { useState } from 'react';
import { calcPrime } from './Algo';

const PrimeForm = () => {
  const [inputNum, setInputNum] = useState('');
  const [answer, setAnswer] = useState('');

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInputNum(e.target.value);
  }

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const num: number = +inputNum
    setAnswer(calcPrime(num));
  }

  return (
    <form id="frmFindHighestPrime" onSubmit={handleSubmit} className="container-fluid nambk-container-input text-center">
      <h2 className="nambk-header">Highest Prime</h2>
      <div className="nambk-header-desc">Lower than Input Number</div>
      <div>
        <input id="inputNumber" name="inputNumber" className="nambk-input" style={{width: '50%'}} 
                type="number" value={inputNum} aria-label="Input Number" autoFocus
                onChange={handleChange} placeholder="Enter a number"/>
        <input id="btnFindHighestPrime" className="nambk-btn nambk-btn-primary" 
                type="submit" value="Go"/>
      </div>
      <div id="result" className="mt-3" aria-readonly style={{minHeight: '2rem',}}>
        {answer}  
      </div>   
    </form>
  );
}
export default PrimeForm;