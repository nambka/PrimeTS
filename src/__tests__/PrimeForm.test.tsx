import React from 'react';
import PrimeForm from '../components/PrimeForm';
import { shallow } from 'enzyme';

describe('PrimeForm: ', () => {
  let wrapper:any;

  beforeEach(() => {
    wrapper = shallow(<PrimeForm />);
  });

  afterEach(() => {
    jest.resetModules();
    jest.resetAllMocks();
    // wrapper.unmount();
  });
  
  it('should render the form and components', () => {
    expect(wrapper.find('form').exists()).toBe(true);
    expect(wrapper.find('input[type="number"]').length).toEqual(1);
    expect(wrapper.find('input[type="submit"]').length).toEqual(1);
    expect(wrapper.find('#result').length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('has a title of "Prime Finder"', () => {
    expect(wrapper.find("h2").text()).toBe("Highest Prime");
    expect(wrapper).toMatchSnapshot();
  });

  it('renders one form component', () => {
    expect(wrapper.find('form')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('includes 02 inputs', () => {
    expect(wrapper.find('input').length).toEqual(2);
  });

  // https://medium.com/@acesmndr/testing-react-functional-components-with-hooks-using-enzyme-f732124d320a
  // Effective Unit Testing of React Components With Jest And Enzyme (TQ): https://xiaoyunyang.github.io/post/effective-unit-testing-of-react-components-with-jest-and-enzyme/
  it('should have proper props for input field', () => {
    expect(wrapper.find('input[type="number"]').props()).toEqual({
      className: 'nambk-input',
      onChange: expect.any(Function),
      placeholder: 'Enter a number',
      type: 'number',
      'aria-label': 'Input Number',
      autoFocus: true,
      id: 'inputNumber',
      name: 'inputNumber',
      style: Object({width: '50%'}),
      value: '',
    });
  });

  it('should have proper props for submit button', () => {
    expect(wrapper.find('input[type="submit"]').props()).toEqual({
      className: 'nambk-btn nambk-btn-primary',
      id: 'btnFindHighestPrime',
      value: 'Go',
      type: 'submit',
    });
  });

  // Ref: https://blog.logrocket.com/getting-started-with-enzyme-for-react-a106b58fc53b/
  // React 17.0.1 error: Cannot read property 'child' of undefined 
  // Enzyme adapter 16 not fully compatible with React 17 
  // => https://github.com/enzymejs/enzyme/issues/2462
  it('should change the value on input', () => {
    wrapper.find('input[type="number"]').simulate('change', { 
      target: { value: 200 } 
    });
    // expect(wrapper.find('input[type="number"]').props().value).toEqual(200);
    expect(wrapper.find('input[type="number"]').prop('value')).toEqual(200);
  });

  it('simulates submit button click event', () => {
    const props = {
      inputNum: 4,
      answer: 3
    }
    const mockSubmit = jest.fn();
    const handleClick = jest.spyOn(React, 'useState');
    wrapper = shallow(<PrimeForm {...props}/>);
    handleClick.mockImplementation((init) => [init, mockSubmit]);
  
    wrapper.find('#btnFindHighestPrime').simulate('click');
    wrapper.update();
    expect(mockSubmit).toBeTruthy();
  });

  it('simulates form submit event', () => {
    const props = {
      inputNum: 4,
      answer: 3
    }
    const fakeEvent = { preventDefault: () => console.log('preventDefault') };
    const mockSubmit = jest.fn();
    const handleSubmit = jest.spyOn(React, 'useState');
    wrapper = shallow(<PrimeForm {...props} onSubmit={handleSubmit}/>);
    handleSubmit.mockImplementation((init) => [init, mockSubmit]);
  
    wrapper.find('form').simulate('submit', fakeEvent);
    expect(mockSubmit).toBeTruthy();
    // expect(mockSubmit).toBeCalledTimes(1);
    // expect(wrapper.find('#result').prop('value')).toEqual(3);
    // expect(wrapper.find('#result').text()).toContain('3');

  });

});