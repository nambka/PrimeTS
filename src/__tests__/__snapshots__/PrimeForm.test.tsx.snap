// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimeForm:  has a title of "Prime Finder" 1`] = `
<form
  className="container-fluid nambk-container-input text-center"
  id="frmFindHighestPrime"
  onSubmit={[Function]}
>
  <h2
    className="nambk-header"
  >
    Highest Prime
  </h2>
  <div
    className="nambk-header-desc"
  >
    Lower than Input Number
  </div>
  <div>
    <input
      aria-label="Input Number"
      autoFocus={true}
      className="nambk-input"
      id="inputNumber"
      name="inputNumber"
      onChange={[Function]}
      placeholder="Enter a number"
      style={
        Object {
          "width": "50%",
        }
      }
      type="number"
      value=""
    />
    <input
      className="nambk-btn nambk-btn-primary"
      id="btnFindHighestPrime"
      type="submit"
      value="Go"
    />
  </div>
  <div
    aria-readonly={true}
    className="mt-3"
    id="result"
    style={
      Object {
        "minHeight": "2rem",
      }
    }
  />
</form>
`;

exports[`PrimeForm:  renders one form component 1`] = `
<form
  className="container-fluid nambk-container-input text-center"
  id="frmFindHighestPrime"
  onSubmit={[Function]}
>
  <h2
    className="nambk-header"
  >
    Highest Prime
  </h2>
  <div
    className="nambk-header-desc"
  >
    Lower than Input Number
  </div>
  <div>
    <input
      aria-label="Input Number"
      autoFocus={true}
      className="nambk-input"
      id="inputNumber"
      name="inputNumber"
      onChange={[Function]}
      placeholder="Enter a number"
      style={
        Object {
          "width": "50%",
        }
      }
      type="number"
      value=""
    />
    <input
      className="nambk-btn nambk-btn-primary"
      id="btnFindHighestPrime"
      type="submit"
      value="Go"
    />
  </div>
  <div
    aria-readonly={true}
    className="mt-3"
    id="result"
    style={
      Object {
        "minHeight": "2rem",
      }
    }
  />
</form>
`;

exports[`PrimeForm:  should render the form and components 1`] = `
<form
  className="container-fluid nambk-container-input text-center"
  id="frmFindHighestPrime"
  onSubmit={[Function]}
>
  <h2
    className="nambk-header"
  >
    Highest Prime
  </h2>
  <div
    className="nambk-header-desc"
  >
    Lower than Input Number
  </div>
  <div>
    <input
      aria-label="Input Number"
      autoFocus={true}
      className="nambk-input"
      id="inputNumber"
      name="inputNumber"
      onChange={[Function]}
      placeholder="Enter a number"
      style={
        Object {
          "width": "50%",
        }
      }
      type="number"
      value=""
    />
    <input
      className="nambk-btn nambk-btn-primary"
      id="btnFindHighestPrime"
      type="submit"
      value="Go"
    />
  </div>
  <div
    aria-readonly={true}
    className="mt-3"
    id="result"
    style={
      Object {
        "minHeight": "2rem",
      }
    }
  />
</form>
`;
